src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java

```diff
@@ -914,13 +917,13 @@ public StructImpl downloadAndExtract(
 
     StructImpl downloadResult = calculateDownloadResult(checksum, downloadedPath);
     try {
-      if (downloadedPath.exists()) {
-        downloadedPath.delete();
+      if (downloadDirectory.exists()) {
+        downloadDirectory.deleteTree();
```

Comment:
In case multiple of these `downloadAndExtract` operation happen in parallel, one thread might try to delete the `tmp` directory, while the other is still busy extracting, which can cause an error.

Could we instead of using a hardcoded `tmp` directory to put the ZIP file in use a securely generated random temporary file name to download the ZIP file to and then in a finally block ensure that we delete it?

We can probably use Java I/O here to create a temporary file:

```
Path tempfile = fileSystem.getPath(File.createTempFile(prefix, suffix, outputPath).getAbsolutePath());
```

Prefix can be anything you like, suffix has to be the actual suffix of the archive (e.g. `.tar.gz` or `.zip`) so that the detection logic for which decompressor to use stays working.

Reply:
Yes, I think this will be better. But just to make sure, I think I will use `Files.createTempDirectory()` to create the `tmp` directory and pass this to the download manager to download the archive in it. The downloaded file is created by the download manager using VFS API.

Reply:
Never mind, I can see now that we can pass the required file path to `downloadManager.download()`.

Reply:
I worked on creating a temporary file using Java I/O but I found that there is some logic used to determine the suffix of the archive from the main url implemented in the DownloadManager. I am not sure if I should do a similar logic to create the temp file or just create a temp directory and pass it to the download manager.

Reply:
This is a good point. It is probably easier to just create a temporary directory, put the file into it with the original filename and then later delete the directory again. I also cannot think of a drawback with that approach, so let's do that. :)